Replace NetworkError::Internal with structured enum variants for calls with literal strings

uthmaniv opened the pull request:
Replace NetworkError::Internal with structured enum variants
- Adds UnsupportedScheme,CorsViolation,ConnectionFailure,Timeout,RedirectError,InvalidMethod,ResourceError,SecurityBlock,MixedContent,CacheError,InvalidPort, LocalDirectoryError, variants in NetworkError enum.
- Refactored the usage of NetworkError::Internal(String) to use the appropriate new variant

Testing: Changes does not require test.
Fixes: https://github.com/servo/servo/issues/36434

uthmaniv committed:
Replace NetworkError::Internal with structured enum variants

Signed-off-by: Uthman Yahaya Baba <uthmanyahayababa@gmail.com>

jdm commented on components/shared/net/lib.rs:
```diff
@@ -933,6 +933,18 @@ pub enum NetworkError {
+    UnsupportedScheme,
+    CorsViolation,
+    ConnectionFailure,
+    Timeout,
```
Timeout seems to be unused.

jdm commented on components/shared/net/lib.rs:
```diff
@@ -933,6 +933,18 @@ pub enum NetworkError {
+    RedirectError,
+    InvalidMethod,
+    ResourceError,
+    SecurityBlock,
```
I'm worried that SecurityBlock covers too many different failures modes and will make it harder to track down the cause of errors. Can we make it more granular?

jdm commented on components/shared/net/lib.rs:
```diff
@@ -933,6 +933,18 @@ pub enum NetworkError {
     /// Crash error, to be converted to Resource::Crash in the HTML parser.
     Crash(String),
+    UnsupportedScheme,
+    CorsViolation,
```
I'm worried that CorsViolation covers too many different failure modes and will be hard to diagnose. Can we make it more granular?

jdm reviewed (changes requested):
Good start!

uthmaniv committed:
Replace NetworkError::CorsViolation and NetworkError::SecurityBlock  with granular enum variants

Signed-off-by: Uthman Yahaya Baba <uthmanyahayababa@gmail.com>

uthmaniv committed:
Replace NetworkError::CorsViolation and NetworkErrorr::SecurityBlock  with granular enum variants

Signed-off-by: Uthman Yahaya Baba <uthmanyahayababa@gmail.com>

jdm commented on components/shared/net/lib.rs:
```diff
@@ -933,6 +933,27 @@ pub enum NetworkError {
+    MixedContent,
+    CacheError,
+    InvalidPort,
+    LocalDirectoryError,
```
The other thing I would like to see here is a new Debug implementation for Network error: https://doc.rust-lang.org/stable/std/fmt/trait.Debug.html

This is where we could move the string values so we don't lose the context from those messages. It would help us expose the actual failure information in https://github.com/servo/servo/blob/a57ab6decaab412f339ece4c79b12aad06979a90/components/script/fetch.rs#L230, which someone asked about on zulip.

jdm replied:
We would match on the enum value and write the specific string to the debug output.

uthmaniv replied:
Ok noted, will make the changes as instructed, thanks

Barry-dE commented:
The PR isn't getting linked to the issue. I think removing "part of" before the issue number fixes this.

uthmaniv committed:
Debug Implementation for NetworkError

Signed-off-by: Uthman Yahaya Baba <uthmanyahayababa@gmail.com>

uthmaniv committed:
Debug Implementation for NetworkError

Signed-off-by: Uthman Yahaya Baba <uthmanyahayababa@gmail.com>

uthmaniv committed:
Replace NetworkError::CorsViolation and NetworkErrorr::SecurityBlock  with granular enum variants

Signed-off-by: Uthman Yahaya Baba <uthmanyahayababa@gmail.com>

jdm commented on components/net/fetch/methods.rs:
```diff
@@ -46,6 +46,7 @@ use crate::protocols::ProtocolRegistry;
 use crate::request_interceptor::RequestInterceptor;
 use crate::subresource_integrity::is_response_integrity_valid;
 
+#[allow(dead_code)]
```
It would be good to continue using this string in the debug implementation for this particular error, rather than marking it as dead code.

uthmaniv replied:
Should I then make CacheError variant accept a string as an argument?

jdm commented on components/net/http_loader.rs:
```diff
@@ -1022,7 +1018,7 @@ pub async fn http_redirect_fetch(
     // Step 7: If request’s redirect count is 20, then return a network error.
     if request.redirect_count >= 20 {
-        return Response::network_error(NetworkError::Internal("Too many redirects".into()));
+        return Response::network_error(NetworkError::RedirectError);
```
We should have a separate error for TooManyRedirects.

jdm replied:
```suggestion
        return Response::network_error(NetworkError::TooManyRedirects);
```

jdm commented on components/net/http_loader.rs:
```diff
@@ -1984,7 +1970,7 @@ async fn http_network_fetch(
     let cancellation_listener = context.cancellation_listener.clone();
     if cancellation_listener.cancelled() {
-        return Response::network_error(NetworkError::Internal("Fetch aborted".into()));
+        return Response::network_error(NetworkError::ConnectionFailure);
```
Huh, this error should be LoadCancelled instead. That variant is new since this code was written.

jdm commented on components/shared/net/lib.rs:
```diff
@@ -933,6 +933,61 @@ pub enum NetworkError {
+                write!(f, "Subresource integrity validation failed")
+            },
+            NetworkError::MixedContent => write!(f, "Blocked as mixed content"),
+            NetworkError::CacheError => write!(f, "Cache error"),
```
This cache error can use `"Couldn't find response in cache"` instead.

jdm reviewed (commented):
This is close! Just a couple more requests for clarity in the new errors.

jdm reviewed (approved):
Let's make sure we use the new variant!

uthmaniv committed:
Resolved merge conflict

Signed-off-by: Uthman Yahaya Baba <uthmanyahayababa@gmail.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

uthmaniv committed:
Resolve Conflicts

Signed-off-by: Usman Yahaya Baba <91813795+uthmaniv@users.noreply.github.com>

TimvdLippe committed:
Address reviewer feedback

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe reviewed (approved):
I have addressed the two remaining comments from Josh in a commit and pushed that it to this PR. Thanks for contributing!

uthmaniv commented:
> I have addressed the two remaining comments from Josh in a commit and pushed that it to this PR. Thanks for contributing!

It's a pleasure.  Thank you

TimvdLippe committed:
script: Fix handling of network errors

Rather than listening all individual errors, the only case
where we shouldn't do anything is `LoadCancelled`. In fact,
we should return early here, since otherwise we go through
the full script_thread to eventually return anyways on
line 1262 since there are no page headers available.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>
